File: day-01a/main.py
import sys

def calculate(filename: str) -> int:
    total = 0
    with open(filename) as f:
        for line in f:
            first: int | None = None
            last: int | None = None
            for char in line:
                if char.isdigit():
                    if first is None:
                        first = int(char)
                    last = int(char)
            assert first is not None, "Oops, Advent of Code is broken!"
            assert last is not None, "Oops, Advent of Code is broken!"
            total += (first * 10) + last
    return total


def main(filename: str) -> None:
    print(calculate(filename))


if __name__ == "__main__":
    main(sys.argv[1])

File: day-01a/main.rs
use std::fs::read_to_string;

fn calculate(filename: &str) -> u32 {
    let mut total = 0;
    for line in read_to_string(filename).unwrap().lines() {
        let mut first = None;
        let mut last = None;
        for char in line.chars() {
            if char.is_ascii_digit() {
                if first.is_none() {
                    first = char.to_digit(10)
                };
                last = char.to_digit(10);
            }
        }
        match (first, last) {
            (Some(f), Some(l)) => {
                let calibration_value = (f * 10) + l;
                total += calibration_value;
            }
            _ => panic!(),
        };
    }
    total
}

fn main() {
    println!("{}", calculate("input.txt"));
}
